### ISSUE TITLE
Suggestion to throw a UserWarning when the user has forgotten .eval() mode during inference

### ISSUE THREAD

**Issue by seva100** [pytorch/pytorch#76383]
### 🚀 The feature, motivation and pitch

It happens really often that during inference time, one forgets to put the model in eval mode `model.eval()` beforehand. When the network has layers with different train/test behavior, such as Dropout / BatchNorm, this can be a source of bugs that are really hard to trace afterwards.

I was wondering if it's possible for PyTorch to throw a warning when the user has forgotten `.eval()` before doing inference. Not sure if this proposal has been raised before.

The only technical caveat is, how should PyTorch detect if the user is inferring the model or not. I am thinking of ways based on `.grad` checking -- e.g. check if a tensor is passed through the network the second time, and no gradient has been calculated for the whole model between the first and the second call. Perhaps another way would be to throw the warning only within `torch.no_grad()` environment.

**Comment by jbschlosser** [comment]
Thanks for the request! When throwing a warning, we generally want to avoid:
* False positives - users get annoyed if there's no way to disable a warning that incorrectly triggers, and it is misleading.
* False negatives - if users start to depend on a warning to tell them they've forgotten something, the lack of a warning can also be misleading.

That said, I'm not aware of a foolproof way to guarantee that the user intends to do inference, as calling `model.eval()` *is* the current way to specify that inference is intended.

Throwing a warning only when in `torch.no_grad()` or `torch.inference_mode()` may be possible, but my understanding is that this won't help some percentage of users who don't use these modes during evaluation - is this correct?

**Comment by seva100** [comment]
Hi @jbschlosser!
Thanks for the feedback. Would do you think of the suggestion to check whether `.grad is None` for all model parameters before feeding a tensor through a model a sufficient number of times (e.g. 10)?

Otherwise, I guess throwing a warning when `torch.no_grad()` or `torch.inference_mode()` might be beneficial, even though, as you say, it does not cover all range of cases or users.

**Comment by jbschlosser** [comment]
> Would do you think of the suggestion to check whether `.grad is None` for all model parameters before feeding a tensor to a model a sufficient number of times (e.g. 10)?

I'd say the "sufficient number of times" is hard to define in a way that fully avoids false positives. It's common in DRL for example to do N forward passes and accumulate grads before doing a single backwards pass. Also, it'd be a possibly non-trivial perf hit to check the `grad` of every model parameter, especially for large models.

**Comment by seva100** [comment]
Ok, I guess the only option then would be to check it within `torch.no_grad()` environment, right?

**Comment by seva100** [comment]
Also, another idea: perhaps torch could show whether the modules like `nn.Dropout` and `nn.BatchNorm2d` are in training mode when the model info is printed. 
For example, right now when the model is printed, torch shows the following:

```python
In [1]: import torch

In [2]: import torch.nn as nn

In [3]: model = nn.Sequential(nn.Dropout())

In [4]: model
Out[4]: 
Sequential(
  (0): Dropout(p=0.5, inplace=False)
)
```

Perhaps it could show the value of the `.training` field for `nn.Dropout`, like:

```python
In [4]: model
Out[4]: 
Sequential(
  (0): Dropout(p=0.5, inplace=False, training=True)
)
```

What do you think of this suggestion? I guess showing the value of `.training` can encourage a simple habit to do `print(model)` more often. Also `inplace` is already being printed out, which is also an auxiliary information.

**Comment by Simranmehta-dev** [comment]
hey @seva100 i want to work on this issue?

**Comment by seva100** [comment]
Hey @Simranmehta-dev, yes certainly -- I'm not a PyTorch developer but if some of the developers can decide it's a good design decision (some of the ones I outlined above or any other way of implementing it), I'd definitely support it!
